File: docs/reference/pfe_auth/frontend/src/services/logAggregator.ts
/**
 * frontend/src/services/logAggregator.ts — Structured logging for comprehensive testing
 *
 * Captures frontend events to:
 * 1. IndexedDB for persistent storage
 * 2. In-memory queue for session analysis
 * 3. Export to JSON for post-test review
 *
 * Event categories:
 * - 'ui': Page navigation, component rendering
 * - 'interaction': Button clicks, form submissions
 * - 'api': API request/response timing and status
 * - 'analytics': User behavior metrics
 */

interface LogEntry {
  timestamp: string;
  eventType: string;
  category: string;
  data: Record<string, any>;
}

class FrontendLogAggregator {
  private logs: LogEntry[] = [];
  private readonly MAX_LOGS = 1000;
  private readonly DB_NAME = "AdaptIQ_Logs";
  private readonly STORE_NAME = "events";
  private db: IDBDatabase | null = null;

  constructor() {
    this.initIndexedDB();
  }

  /**
   * Initialize IndexedDB for persistent storage
   */
  private initIndexedDB(): void {
    const request = indexedDB.open(this.DB_NAME, 1);

    request.onerror = () => {
      console.warn("Failed to initialize IndexedDB for logging");
    };

    request.onsuccess = (event) => {
      this.db = (event.target as IDBOpenDBRequest).result;
    };

    request.onupgradeneeded = (event) => {
      const db = (event.target as IDBOpenDBRequest).result;
      if (!db.objectStoreNames.contains(this.STORE_NAME)) {
        db.createObjectStore(this.STORE_NAME, { autoIncrement: true });
      }
    };
  }

  /**
   * Log a structured event
   */
  private logEvent(
    eventType: string,
    category: string,
    data: Record<string, any>
  ): void {
    const entry: LogEntry = {
      timestamp: new Date().toISOString(),
      eventType,
      category,
      data,
    };

    // Store in memory
    this.logs.push(entry);
    if (this.logs.length > this.MAX_LOGS) {
      this.logs.shift(); // Remove oldest entry
    }

    // Store in IndexedDB asynchronously
    if (this.db) {
      const store = this.db
        .transaction(this.STORE_NAME, "readwrite")
        .objectStore(this.STORE_NAME);
      store.add(entry);
    }

    // Log to console in development
    if (process.env.NODE_ENV === "development") {
      console.log(`[${category.toUpperCase()}] ${eventType}:`, data);
    }
  }

  /**
   * Log page view
   */
  logPageView(pageName: string): void {
    this.logEvent("page_view", "ui", {
      page: pageName,
      url: window.location.href,
    });
  }

  /**
   * Log user action (click, form submission, etc.)
   */
  logUserAction(
    actionType: string,
    data: Record<string, any> = {}
  ): void {
    this.logEvent("user_action", "interaction", {
      action: actionType,
      ...data,
    });
  }

  /**
   * Log API call with timing and status
   */
  logApiCall(
    endpoint: string,
    method: string,
    status: number,
    durationMs: number
  ): void {
    this.logEvent("api_call", "api", {
      endpoint,
      method,
      status,
      durationMs,
      timestamp: new Date().toISOString(),
    });
  }

  /**
   * Log session event (start, answer, etc.)
   */
  logSessionEvent(
    sessionId: string,
    eventType: string,
    data: Record<string, any> = {}
  ): void {
    this.logEvent(eventType, "session", {
      sessionId,
      ...data,
    });
  }

  /**
   * Log question shown to user
   */
  logQuestionShown(
    sessionId: string,
    questionId: string,
    difficulty: number
  ): void {
    this.logEvent("question_shown", "session", {
      sessionId,
      questionId,
      difficulty,
      userAgent: navigator.userAgent,
    });
  }

  /**
   * Log answer submission
   */
  logAnswerSubmitted(
    sessionId: string,
    questionId: string,
    selectedIndex: number,
    timeTakenMs: number
  ): void {
    this.logEvent("answer_submitted", "session", {
      sessionId,
      questionId,
      selectedIndex,
      timeTakenMs,
    });
  }

  /**
   * Log session completion
   */
  logSessionEnd(
    sessionId: string,
    topic: string,
    questionsAnswered: number,
    correctCount: number
  ): void {
    const accuracy =
      questionsAnswered > 0 ? (correctCount / questionsAnswered) * 100 : 0;
    this.logEvent("session_end", "session", {
      sessionId,
      topic,
      questionsAnswered,
      correctCount,
      accuracy: Math.round(accuracy * 100) / 100,
    });
  }

  /**
   * Log rank change in challenge room
   */
  logRankChange(oldRank: string, newRank: string): void {
    this.logEvent("rank_changed", "session", {
      oldRank,
      newRank,
      timestamp: new Date().toISOString(),
    });
  }

  /**
   * Get all logged events
   */
  getLogs(): LogEntry[] {
    return [...this.logs];
  }

  /**
   * Get recent logs (last N entries)
   */
  getRecentLogs(count: number): LogEntry[] {
    return this.logs.slice(-count);
  }

  /**
   * Get logs by category
   */
  getLogsByCategory(category: string): LogEntry[] {
    return this.logs.filter((log) => log.category === category);
  }

  /**
   * Get analytics summary
   */
  getStats(): Record<string, any> {
    const eventCounts: Record<string, number> = {};
    const categoryCounts: Record<string, number> = {};

    this.logs.forEach((log) => {
      eventCounts[log.eventType] = (eventCounts[log.eventType] || 0) + 1;
      categoryCounts[log.category] = (categoryCounts[log.category] || 0) + 1;
    });

    return {
      totalEvents: this.logs.length,
      eventTypes: eventCounts,
      categories: categoryCounts,
      oldestEvent: this.logs[0]?.timestamp || null,
      newestEvent: this.logs[this.logs.length - 1]?.timestamp || null,
    };
  }

  /**
   * Export logs as JSON
   */
  exportAsJson(): string {
    return JSON.stringify(this.logs, null, 2);
  }

  /**
   * Download logs as file
   */
  downloadLogs(filename: string = "adaptiq_logs.json"): void {
    const json = this.exportAsJson();
    const blob = new Blob([json], { type: "application/json" });
    const url = URL.createObjectURL(blob);
    const a = document.createElement("a");
    a.href = url;
    a.download = filename;
    a.click();
    URL.revokeObjectURL(url);
  }

  /**
   * Clear all logs
   */
  clearLogs(): void {
    this.logs = [];
    if (this.db) {
      this.db
        .transaction(this.STORE_NAME, "readwrite")
        .objectStore(this.STORE_NAME)
        .clear();
    }
  }

  /**
   * Clear logs older than specified hours
   */
  clearOldLogs(hoursOld: number): void {
    const cutoff = new Date(Date.now() - hoursOld * 60 * 60 * 1000);
    this.logs = this.logs.filter((log) => new Date(log.timestamp) > cutoff);
  }
}

// Global instance
let loggerInstance: FrontendLogAggregator | null = null;

export function getLogAggregator(): FrontendLogAggregator {
  if (!loggerInstance) {
    loggerInstance = new FrontendLogAggregator();
  }
  return loggerInstance;
}

// Expose to window for browser console debugging
if (typeof window !== "undefined") {
  (window as any).logAggregator = getLogAggregator();
}

export default FrontendLogAggregator;

File: docs/reference/pfe_auth/backend/services/log_aggregator.py
"""
services/log_aggregator.py — Structured logging for comprehensive testing.

Captures all critical events to both:
1. JSON files (backend/logs/test_session_*.json) for easy review
2. PostgreSQL test_logs table for queryable analysis

Event categories:
- 'irt': IRT calculations (theta updates, difficulty selection)
- 'cache': Cache hit/miss, invalidation
- 'session': Session lifecycle (start, end, state changes)
- 'api': API calls and responses
- 'database': Database operations
- 'ui': Frontend interactions (logged separately in TypeScript)
- 'rank': Challenge room rank changes
"""

import json
import logging
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Optional
from uuid import UUID

logger = logging.getLogger(__name__)


class LogAggregator:
    """Collects structured logs for testing and debugging."""

    def __init__(self, logs_dir: str = "backend/logs"):
        """Initialize log aggregator with file output directory."""
        self.logs_dir = Path(logs_dir)
        self.logs_dir.mkdir(parents=True, exist_ok=True)
        self.session_file = None
        self.session_logs = []

    def log_event(
        self,
        event_type: str,
        category: str,
        data: dict,
        user_id: Optional[UUID] = None,
        session_id: Optional[UUID] = None,
    ) -> None:
        """
        Log a structured event.

        Args:
            event_type: Event type (e.g., 'theta_update', 'cache_hit')
            category: Category (irt, cache, session, api, database, rank)
            data: Event-specific data as dict
            user_id: Associated user (optional)
            session_id: Associated session (optional)
        """
        timestamp = datetime.now(timezone.utc).isoformat()

        log_entry = {
            "timestamp": timestamp,
            "event_type": event_type,
            "category": category,
            "user_id": str(user_id) if user_id else None,
            "session_id": str(session_id) if session_id else None,
            "data": data,
        }

        # Store in memory for session file export
        self.session_logs.append(log_entry)

        # Log to console in development
        logger.info(f"[{category.upper()}] {event_type}: {data}")

    def log_irt_update(
        self,
        user_id: UUID,
        concept_id: UUID,
        old_theta: float,
        new_theta: float,
        p_correct: float,
        learning_rate: float = 0.3,
        question_id: Optional[UUID] = None,
    ) -> None:
        """Log IRT theta update."""
        self.log_event(
            event_type="theta_update",
            category="irt",
            data={
                "concept_id": str(concept_id),
                "old_theta": round(old_theta, 3),
                "new_theta": round(new_theta, 3),
                "theta_change": round(new_theta - old_theta, 3),
                "p_correct": round(p_correct, 3),
                "learning_rate": learning_rate,
                "question_id": str(question_id) if question_id else None,
            },
            user_id=user_id,
        )

    def log_difficulty_selection(
        self,
        user_id: UUID,
        user_theta: float,
        zpd_range: tuple[float, float],
        selected_beta: float,
        question_id: UUID,
    ) -> None:
        """Log difficulty selection for next question."""
        self.log_event(
            event_type="difficulty_selected",
            category="irt",
            data={
                "user_theta": round(user_theta, 3),
                "zpd_min": round(zpd_range[0], 3),
                "zpd_max": round(zpd_range[1], 3),
                "selected_beta": round(selected_beta, 3),
                "question_id": str(question_id),
            },
            user_id=user_id,
        )

    def log_session_start(
        self,
        user_id: UUID,
        session_id: UUID,
        topic: str,
        initial_theta: float,
    ) -> None:
        """Log session start."""
        self.log_event(
            event_type="session_start",
            category="session",
            data={
                "topic": topic,
                "initial_theta": round(initial_theta, 3),
            },
            user_id=user_id,
            session_id=session_id,
        )

    def log_session_end(
        self,
        user_id: UUID,
        session_id: UUID,
        questions_answered: int,
        correct_count: int,
        final_theta: float,
    ) -> None:
        """Log session completion."""
        accuracy = correct_count / questions_answered if questions_answered > 0 else 0
        self.log_event(
            event_type="session_end",
            category="session",
            data={
                "questions_answered": questions_answered,
                "correct_count": correct_count,
                "accuracy": round(accuracy, 3),
                "final_theta": round(final_theta, 3),
            },
            user_id=user_id,
            session_id=session_id,
        )

    def log_question_shown(
        self,
        user_id: UUID,
        session_id: UUID,
        question_id: UUID,
        user_theta: float,
        question_beta: float,
        expected_p_correct: float,
    ) -> None:
        """Log question presentation."""
        self.log_event(
            event_type="question_shown",
            category="session",
            data={
                "question_id": str(question_id),
                "user_theta": round(user_theta, 3),
                "question_beta": round(question_beta, 3),
                "expected_p_correct": round(expected_p_correct, 3),
            },
            user_id=user_id,
            session_id=session_id,
        )

    def log_answer_submitted(
        self,
        user_id: UUID,
        session_id: UUID,
        question_id: UUID,
        was_correct: bool,
        time_taken_ms: int,
    ) -> None:
        """Log answer submission."""
        self.log_event(
            event_type="answer_submitted",
            category="session",
            data={
                "question_id": str(question_id),
                "was_correct": was_correct,
                "time_taken_ms": time_taken_ms,
            },
            user_id=user_id,
            session_id=session_id,
        )

    def log_cache_operation(
        self,
        operation: str,
        key: str,
        hit: bool,
        latency_ms: Optional[float] = None,
    ) -> None:
        """Log cache hit/miss/operation."""
        self.log_event(
            event_type=f"cache_{operation}",
            category="cache",
            data={
                "key": key,
                "hit": hit,
                "latency_ms": latency_ms,
            },
        )

    def log_rank_change(
        self,
        user_id: UUID,
        old_rank: str,
        new_rank: str,
        reason: str,
    ) -> None:
        """Log challenge room rank change."""
        self.log_event(
            event_type="rank_changed",
            category="rank",
            data={
                "old_rank": old_rank,
                "new_rank": new_rank,
                "reason": reason,
            },
            user_id=user_id,
        )

    def log_match_end(
        self,
        user_id: UUID,
        match_id: UUID,
        rank: str,
        questions_answered: int,
        correct_count: int,
        elo_change: float,
    ) -> None:
        """Log challenge match completion."""
        accuracy = correct_count / questions_answered if questions_answered > 0 else 0
        self.log_event(
            event_type="match_end",
            category="rank",
            data={
                "match_id": str(match_id),
                "rank": rank,
                "questions_answered": questions_answered,
                "correct_count": correct_count,
                "accuracy": round(accuracy, 3),
                "elo_change": round(elo_change, 3),
            },
            user_id=user_id,
        )

    def export_session_logs(self, test_name: str = "test_session") -> Path:
        """
        Export accumulated logs to JSON file.

        Returns: Path to exported file
        """
        timestamp = datetime.now(timezone.utc).strftime("%Y%m%d_%H%M%S")
        filename = self.logs_dir / f"{test_name}_{timestamp}.json"

        with open(filename, "w") as f:
            json.dump(self.session_logs, f, indent=2)

        logger.info(f"Exported {len(self.session_logs)} log entries to {filename}")
        return filename

    def get_stats(self) -> dict:
        """Get summary statistics about logged events."""
        event_counts = {}
        category_counts = {}

        for entry in self.session_logs:
            event_type = entry["event_type"]
            category = entry["category"]

            event_counts[event_type] = event_counts.get(event_type, 0) + 1
            category_counts[category] = category_counts.get(category, 0) + 1

        return {
            "total_events": len(self.session_logs),
            "event_types": event_counts,
            "categories": category_counts,
        }


# Global instance for convenient access
_aggregator = None


def get_aggregator() -> LogAggregator:
    """Get or create the global log aggregator."""
    global _aggregator
    if _aggregator is None:
        _aggregator = LogAggregator()
    return _aggregator
